rclpy/src/rclpy/_rclpy_action.c

```diff
@@ -0,0 +1,929 @@
+ *
+ *     For a rcl_action_server_t:
+ *       (is_goal_request_ready,
+ *        is_cancel_request_ready,
+ *        is_result_request_ready,
+ *        is_goal_expired)
+ */
+static PyObject *
+rclpy_action_wait_set_is_ready(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pyentity;
+  PyObject * pywait_set;
+
+  if (!PyArg_ParseTuple(args, "OO", &pyentity, &pywait_set)) {
+    return NULL;
+  }
+
+  rcl_wait_set_t * wait_set = (rcl_wait_set_t *)PyCapsule_GetPointer(pywait_set, "rcl_wait_set_t");
+  if (!wait_set) {
+    return NULL;
+  }
+
+  if (PyCapsule_IsValid(pyentity, "rcl_action_client_t")) {
+    rcl_action_client_t * action_client =
+      (rcl_action_client_t *)PyCapsule_GetPointer(pyentity, "rcl_action_client_t");
+    bool is_feedback_ready = false;
+    bool is_status_ready = false;
+    bool is_goal_response_ready = false;
+    bool is_cancel_response_ready = false;
+    bool is_result_response_ready = false;
+    rcl_ret_t ret = rcl_action_client_wait_set_get_entities_ready(
+      wait_set,
+      action_client,
+      &is_feedback_ready,
+      &is_status_ready,
+      &is_goal_response_ready,
+      &is_cancel_response_ready,
+      &is_result_response_ready);
+    if (RCL_RET_OK != ret) {
+      PyErr_Format(PyExc_RuntimeError,
+        "Failed to get number of ready entities for action client: %s",
+        rcl_get_error_string().str);
+      rcl_reset_error();
+      return NULL;
+    }
+
+    PyObject * result_tuple = PyTuple_Pack(5,
+        PyBool_FromLong(is_feedback_ready),
+        PyBool_FromLong(is_status_ready),
+        PyBool_FromLong(is_goal_response_ready),
+        PyBool_FromLong(is_cancel_response_ready),
+        PyBool_FromLong(is_result_response_ready));
+    if (!result_tuple) {
+      return PyErr_Format(PyExc_RuntimeError, "Failed to create result tuple for action client");
+    }
+    return result_tuple;
+  } else if (PyCapsule_IsValid(pyentity, "rcl_action_server_t")) {
+    rcl_action_server_t * action_server =
+      (rcl_action_server_t *)PyCapsule_GetPointer(pyentity, "rcl_action_server_t");
+    bool is_goal_request_ready = false;
+    bool is_cancel_request_ready = false;
+    bool is_result_request_ready = false;
+    bool is_goal_expired = false;
+    rcl_ret_t ret = rcl_action_server_wait_set_get_entities_ready(
+      wait_set,
+      action_server,
+      &is_goal_request_ready,
+      &is_cancel_request_ready,
+      &is_result_request_ready,
+      &is_goal_expired);
+    if (RCL_RET_OK != ret) {
+      PyErr_Format(PyExc_RuntimeError,
+        "Failed to get number of ready entities for action server: %s",
+        rcl_get_error_string().str);
+      rcl_reset_error();
+      return NULL;
+    }
+
+    PyObject * result_tuple = PyTuple_Pack(4,
+        PyBool_FromLong(is_goal_request_ready),
+        PyBool_FromLong(is_cancel_request_ready),
+        PyBool_FromLong(is_result_request_ready),
+        PyBool_FromLong(is_goal_expired));
```

Comment:
memory is leaked here too if `PyTuple_Pack` fails

Reply:
Fixed 34db026